crypto/tbs/src/lib.rs

```diff
@@ -174,40 +175,49 @@ pub fn sign_blinded_msg(msg: BlindedMessage, sks: SecretKeyShare) -> BlindedSign
     BlindedSignatureShare(sig.to_affine())
 }
 
-/// Combines a sufficient amount of valid blinded signature shares to a blinded
-/// signature. The responsibility of verifying the supplied shares lies with the
-/// caller.
-///
-/// * `sig_shares`: an iterator yielding pairs of key indices and signature
-///   shares from said key
-/// * `threshold`: number of shares needed to combine a signature
-///
+/// Combines the exact threshold of valid blinded signature shares to a blinded
+/// signature. The responsibility of verifying the shares lies and supplying
+/// exactly the necessary threshold of shares lies with the caller.
 /// # Panics
-/// If the amount of shares supplied is less than the necessary amount
-pub fn combine_valid_shares<I>(sig_shares: I, threshold: usize) -> BlindedSignature
-where
-    I: IntoIterator<Item = (usize, BlindedSignatureShare)>,
-    I::IntoIter: Clone + ExactSizeIterator,
-{
-    let points = sig_shares
-        .into_iter()
-        .take(threshold)
-        .map(|(idx, share)| {
-            let x = Scalar::from((idx as u64) + 1);
-            let y = share.0.into();
-            (x, y)
-        })
-        .collect::<Vec<(Scalar, G1Projective)>>();
-    if points.len() < threshold {
-        panic!("Not enough signature shares");
-    }
+/// If the amount of shares supplied is less than two
+pub fn aggregate_signature_shares(
+    shares: &BTreeMap<u64, BlindedSignatureShare>,
+) -> BlindedSignature {
+    BlindedSignature(
+        lagrange_multiplier(shares.keys().cloned().map(Scalar::from).collect())
+            .into_iter()
+            .zip(shares.values())
+            .map(|(lm, y)| lm * y.0)
+            .reduce(|a, b| a + b)
```

Comment:
Please use more descriptive names, this ain't a math paper :stuck_out_tongue_winking_eye:, e.g. `lagrange_m` instead of `lm` (whatever m stands for xD)

Reply:
IMO if the scope of a symbol is tiny enough, it's kind of OK to go supa-short-names, as it helps readability.

Also [`reduce(Add::add)`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3e921efa3500d42945a5aff8b520a8f7) works and avoids naming anything. 

And isn't the whole thing just a `.sum()`?

Reply:
sum does not work because in case of an empty iterator it returns the zero type but we want to panic